internal/app/machined/pkg/controllers/perf/perf.go

```diff
@@ -0,0 +1,112 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package perf
+
+import (
+	"context"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/prometheus/procfs"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/resources/perf"
+)
+
+const updateInterval = time.Second * 30
+
+// StatsController manages v1alpha1.Stats which is the current snaphot of the machine CPU and Memory consumption.
+type StatsController struct{}
+
+// Name implements controller.StatsController interface.
+func (ctrl *StatsController) Name() string {
+	return "perf.StatsController"
+}
+
+// Inputs implements controller.StatsController interface.
+func (ctrl *StatsController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.StatsController interface.
+func (ctrl *StatsController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: perf.CPUType,
+			Kind: controller.OutputExclusive,
+		},
+		{
+			Type: perf.MemoryType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.StatsController interface.
+func (ctrl *StatsController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	ticker := time.NewTicker(updateInterval)
+
+	defer ticker.Stop()
+
+	var (
+		fs          procfs.FS
+		procfsReady bool
+		err         error
+	)
+
+	for {
+		if !procfsReady {
+			fs, err = procfs.NewDefaultFS()
+			if err != nil {
+				return err
+			}
+		}
+
+		if err := ctrl.updateMemory(ctx, r, &fs); err != nil {
+			return err
+		}
+
+		if err := ctrl.updateCPU(ctx, r, &fs); err != nil {
+			return err
+		}
+
+		select {
```

Comment:
if you add here `case <-r.EventCh()` and move it to the beginning of `for`, it will work correctly - runtime delivers initial reconcile event on controller start (or restart)